File: sdks/python/sealstack/_http.py
"""Internal async HTTP client with retry, full jitter, and redaction.

Mirrors the TS SDK's HttpClient one-for-one. async-first; the sync
facade lives in `client.py` and wraps each async method with
`asyncio.run`.
"""

from __future__ import annotations

import asyncio
import random
from dataclasses import dataclass
from typing import Any, Callable

import httpx

from .errors import (
    BackendError,
    RateLimitedError,
    from_wire_error,
)

REDACTED_HEADERS = frozenset({
    "authorization",
    "cookie",
    "x-api-key",
    "x-sealstack-user", "x-sealstack-tenant", "x-sealstack-roles",
    "x-cfg-user", "x-cfg-tenant", "x-cfg-roles",
})


# Per-request headers. Static dict for the common case; a callable
# returning a dict for token-rotating clients
# (`SealStack.bearer(token=lambda: fetch_jwt())`), which would otherwise
# bind to a single token at client construction.
HeadersSource = dict[str, str] | Callable[[], dict[str, str]]


@dataclass
class HttpClientOptions:
    base_url: str
    headers: HeadersSource
    timeout_s: float
    retry_attempts: int
    retry_initial_backoff_ms: int
    debug: Callable[[str], None] | None = None


class HttpClient:
    def __init__(self, opts: HttpClientOptions) -> None:
        self._opts = opts
        # Construction-time headers are intentionally absent: httpx merges
        # client-level headers with per-request headers, so any
        # constructor-set headers would silently persist alongside the
        # factory-resolved auth header and partially re-introduce the
        # stale-token bug. _resolve_headers() must remain the sole source.
        self._client = httpx.AsyncClient(
            base_url=opts.base_url.rstrip("/"),
            timeout=opts.timeout_s,
        )

    def _resolve_headers(self) -> dict[str, str]:
        h = self._opts.headers
        return h() if callable(h) else h

    async def aclose(self) -> None:
        await self._client.aclose()

    def _log_request_for_test(self, method: str, path: str) -> None:
        self._log_request(method, path, self._resolve_headers())

    async def request(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        no_retry: bool = False,
        timeout_s: float | None = None,
    ) -> Any:
        max_attempts = 1 if no_retry else self._opts.retry_attempts + 1
        last_error: Exception | None = None

        for attempt in range(1, max_attempts + 1):
            try:
                return await self._attempt(method, path, body, timeout_s)
            except Exception as e:
                last_error = e
                if not self._should_retry(e, attempt, max_attempts):
                    raise
                await self._sleep(self._backoff_ms(attempt, e) / 1000.0)

        assert last_error is not None
        raise last_error

    async def _attempt(
        self, method: str, path: str, body: Any, timeout_s: float | None
    ) -> Any:
        # Resolve headers exactly once per attempt: a token-rotation factory
        # backed by a remote JWT fetch must not fire twice (once for the log,
        # once for the request) and the log must show the same token that
        # actually went out on the wire.
        headers = self._resolve_headers()
        self._log_request(method, path, headers)
        timeout = timeout_s if timeout_s is not None else self._opts.timeout_s
        resp = await self._client.request(
            method, path, json=body, headers=headers, timeout=timeout
        )
        headers = {k.lower(): v for k, v in resp.headers.items()}
        env = resp.json() if resp.content else {"data": None, "error": None}

        if resp.status_code >= 400 or env.get("error"):
            self._log_error_response(resp.status_code, headers, resp.text)
            wire = env.get("error") or {"code": "backend", "message": f"HTTP {resp.status_code}"}
            raise from_wire_error(wire, headers=headers)
        if env.get("data") is None:
            raise BackendError(
                "response envelope missing data",
                request_id=headers.get("x-request-id", "unknown"),
            )
        return env["data"]

    def _should_retry(self, e: Exception, attempt: int, max_attempts: int) -> bool:
        if attempt >= max_attempts:
            return False
        if isinstance(e, (RateLimitedError, BackendError)):
            return True
        if isinstance(e, httpx.TransportError):  # network errors
            return True
        return False

    def _backoff_ms(self, attempt: int, e: Exception) -> float:
        if isinstance(e, RateLimitedError) and e.retry_after is not None and e.retry_after >= 0:
            return e.retry_after * 1000.0
        base = self._opts.retry_initial_backoff_ms * (2 ** (attempt - 1))
        # Full jitter: uniform random in [0, base * 1.25].
        return random.uniform(0.0, base * 1.25)

    async def _sleep(self, seconds: float) -> None:
        # asyncio.sleep is naturally cancellable; cancellation propagates
        # through the retry loop without explicit handling.
        await asyncio.sleep(seconds)

    def _log_request(
        self, method: str, path: str, headers: dict[str, str]
    ) -> None:
        if self._opts.debug is None:
            return
        self._opts.debug(f"→ {method} {path} headers={self._redact_headers(headers)}")

    def _log_error_response(
        self, status: int, headers: dict[str, str], body: str
    ) -> None:
        if self._opts.debug is None:
            return
        self._opts.debug(
            f"← {status} headers={self._redact_headers(headers)} body={body}"
        )

    @staticmethod
    def _redact_headers(h: dict[str, str]) -> dict[str, str]:
        return {
            k.lower(): ("<redacted>" if k.lower() in REDACTED_HEADERS else v)
            for k, v in h.items()
        }

File: sdks/typescript/src/http.ts
import { fromWireError, SealStackError, BackendError } from "./errors.js";

/** Headers redacted from debug logs (case-insensitive). */
export const REDACTED_HEADERS = new Set([
  "authorization",
  "cookie",
  "x-api-key",
  "x-sealstack-user",
  "x-sealstack-tenant",
  "x-sealstack-roles",
  "x-cfg-user",
  "x-cfg-tenant",
  "x-cfg-roles",
]);

/**
 * Per-request headers. Static dict for the common case; a factory `() => dict`
 * for token-rotating clients (`SealStack.bearer({ token: () => fetch_jwt() })`),
 * which would otherwise bind to a single token at client construction.
 */
export type HeadersSource = Record<string, string> | (() => Record<string, string>);

export interface HttpClientOptions {
  baseUrl: string;
  headers: HeadersSource;
  timeoutMs: number;
  retryAttempts: number;
  retryInitialBackoffMs: number;
  debug?: (msg: string) => void;
}

export interface RequestOptions {
  method: "GET" | "POST";
  path: string;
  body?: unknown;
  signal?: AbortSignal;
  /** Per-call override of the client default. */
  timeoutMs?: number;
  /** Skip retry policy (admin namespace uses this). */
  noRetry?: boolean;
}

interface Envelope<T> {
  data: T | null;
  error: { code: string; message: string } | null;
}

export class HttpClient {
  constructor(private opts: HttpClientOptions) {}

  private resolveHeaders(): Record<string, string> {
    return typeof this.opts.headers === "function" ? this.opts.headers() : this.opts.headers;
  }

  /** Public test hook for the redaction logic; not for production callers. */
  logRequestForTest(req: { method: string; path: string }): void {
    this.logRequest(req, this.resolveHeaders());
  }

  async request<T>(req: RequestOptions): Promise<T> {
    const maxAttempts = req.noRetry ? 1 : this.opts.retryAttempts + 1;
    const timeoutMs = req.timeoutMs ?? this.opts.timeoutMs;
    let lastError: unknown;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        return await this.attempt<T>(req, timeoutMs);
      } catch (e) {
        lastError = e;
        if (req.signal?.aborted) throw e;
        if (!this.shouldRetry(e, attempt, maxAttempts)) throw e;
        await this.sleepWithCancel(this.backoffMs(attempt, e), req.signal);
      }
    }
    throw lastError;
  }

  private async attempt<T>(req: RequestOptions, timeoutMs: number): Promise<T> {
    // Resolve headers exactly once per attempt: a token-rotation factory
    // backed by a remote JWT fetch must not fire twice (once for the log,
    // once for the request) and the log must show the same token that
    // actually went out on the wire.
    const reqHeaders = this.resolveHeaders();
    this.logRequest(req, reqHeaders);
    const url = `${this.opts.baseUrl}${req.path}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    if (req.signal) req.signal.addEventListener("abort", () => controller.abort(), { once: true });

    let res: Response;
    try {
      res = await fetch(url, {
        method: req.method,
        headers: { "content-type": "application/json", ...reqHeaders },
        body: req.body == null ? undefined : JSON.stringify(req.body),
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timer);
    }

    const headers = headersToObject(res.headers);
    const text = await res.text();
    const env: Envelope<T> = text ? JSON.parse(text) : { data: null, error: null };

    if (res.status >= 400 || env.error) {
      this.logErrorResponse(res.status, headers, text);
      const wireErr = env.error ?? { code: "backend", message: `HTTP ${res.status}` };
      throw fromWireError(wireErr, { headers });
    }
    if (env.data == null) {
      throw new BackendError("response envelope missing data", headers["x-request-id"] ?? "unknown");
    }
    return env.data;
  }

  private shouldRetry(e: unknown, attempt: number, max: number): boolean {
    if (attempt >= max) return false;
    if (e instanceof SealStackError) {
      // Retriable: rate_limited (429) and backend (5xx). Per spec §9.1.
      return e.constructor.name === "RateLimitedError"
          || e.constructor.name === "BackendError";
    }
    // Network/abort errors retry.
    return e instanceof TypeError; // fetch network failure
  }

  private backoffMs(attempt: number, e: unknown): number {
    // Honor Retry-After on RateLimitedError if present.
    if (e instanceof SealStackError && e.constructor.name === "RateLimitedError") {
      const ra = (e as { retryAfter?: number }).retryAfter;
      if (ra != null && ra >= 0) return ra * 1000;
    }
    const base = this.opts.retryInitialBackoffMs * 2 ** (attempt - 1);
    // Full jitter: uniform random in [0, base * 1.25].
    return Math.random() * base * 1.25;
  }

  private async sleepWithCancel(ms: number, signal: AbortSignal | undefined): Promise<void> {
    return new Promise((resolve, reject) => {
      const t = setTimeout(resolve, ms);
      signal?.addEventListener("abort", () => {
        clearTimeout(t);
        reject(new Error("aborted"));
      }, { once: true });
    });
  }

  private logRequest(
    req: { method: string; path: string },
    headers: Record<string, string>,
  ): void {
    if (!this.opts.debug) return;
    this.opts.debug(`→ ${req.method} ${req.path} headers=${JSON.stringify(redactHeaders(headers))}`);
  }

  private logErrorResponse(status: number, headers: Record<string, string>, body: string): void {
    if (!this.opts.debug) return;
    this.opts.debug(`← ${status} headers=${JSON.stringify(redactHeaders(headers))} body=${body}`);
  }
}

function redactHeaders(h: Record<string, string>): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [k, v] of Object.entries(h)) {
    out[k.toLowerCase()] = REDACTED_HEADERS.has(k.toLowerCase()) ? "<redacted>" : v;
  }
  return out;
}

function headersToObject(h: Headers): Record<string, string> {
  const out: Record<string, string> = {};
  h.forEach((v, k) => { out[k.toLowerCase()] = v; });
  return out;
}
